### DISCUSSION TITLE
https for embedded use case (certificates)

### DISCUSSION THREAD

**Discussion by nicolasb565** [babelouest/ulfius#228] [Q&A] [answered]
Hi,

I am wondering how you generate the certificate and private key for ulfius_start_secure_framework().

I tried a few examples from libmicrohttpd documentation but I did not manage to make it work.

Thanks!

**Comment by babelouest** [comment, marked as answer]
Hello,

For tests purpose, you can generate a self-signed certificate using openssl (see [example](https://stackoverflow.com/a/41366949/12247563)) or GnuTLS' certtool:

```shell
$ # example with openssl
$ openssl req -x509 -newkey rsa:4096 -sha256 -days 3650 -nodes \
  -keyout localhost.key -out localhost.crt -subj "/CN=localhost"
$
$ # example with certtool
$ certtool --generate-privkey --outfile localhost.key --sec-param High
$ certtool --generate-self-signed --load-privkey localhost.key --outfile localhost.crt
```

Then, you can use the content of the files `localhost.key` and `localhost.crt` as parameters `key_pem` and `cert_pem` in the function `ulfius_start_secure_framework`:

```C
const char key_pem[] =
"-----BEGIN RSA PRIVATE KEY-----\
MIIG4wIBAA[...]LolR+U2f7QAiM1K+x\
-----END RSA PRIVATE KEY-----";
const char cert_pem[] =
"-----BEGIN CERTIFICATE-----
MIIEJDCCAoygAw[...]AAJx6S6cOEEM1\
+L6JIqXTCzQ=
-----END CERTIFICATE-----";
ulfius_start_secure_framework(&instance, key_pem, cert_pem);
```
